### PR TITLE
PEtab SciML test suite - test_net and (some) test_ude cases

### PR DESCRIPTION
**Pull request by BSnelling**
Implements support for petab sciml test suite: https://github.com/sebapersson/petab_sciml_testsuite/tree/main

Includes support for all `test_net` cases and a subset of `test_ude`. Test cases with frozen layers and networks in the observable formulae are not yet implemented.

### CODE REVIEW

File: python/sdist/amici/jax/petab.py [hunk 1]
```diff
@@ -524,28 +527,51 @@ def _get_nominal_parameter_values(
...
+            if (net := pname.split("_")[0]) in model.nns:
                 to_set = []
                 nn = model_pars[net]
-                try:
-                    value = float(row[petab.NOMINAL_VALUE])
-                except ValueError:
-                    value = par_arrays[row[petab.NOMINAL_VALUE]]
+                scalar = True
+
+                # ?? When value is NaN do we want to go to par_arrays ??
```
Comment (FFroehlich) [inline python/sdist/amici/jax/petab.py]: Modern IDEs should support [Codetags](https://peps.python.org/pep-0350/) which can make it easier to track these comments

File: python/sdist/amici/jax/petab.py [hunk 2]
```diff
@@ -588,6 +614,38 @@ def _get_nominal_parameter_values(
...
+                                    [
+                                        (
+                                            k,
+                                            jnp.array(
+                                                nn_input_arrays[net_id][input][
+                                                    k
+                                                ][:],
+                                                dtype=jnp.float64,
+                                            ),
+                                        )  # ?? hardcoded dtype not ideal ?? could infer from env somehow ??
```
Comment (FFroehlich) [inline python/sdist/amici/jax/petab.py]: is it really necessary to set `dtype` here? Usually jax infers float precision from https://docs.jax.dev/en/latest/config_options.html. Might be necessary to cast this as numpy array first if conversion from hdf5 is the problem

File: python/sdist/amici/jax/petab.py [hunk 3]
```diff
@@ -709,15 +780,70 @@ def _eval_nn(self, output_par: str):
...
+
+        hybridization_parameter_map = dict(
+            [
+                (petab_id, self._hybridization_df.loc[petab_id, "targetValue"])
+                for petab_id in model_id_map.values()
+                if petab_id in set(self._hybridization_df.index)
+            ]
+        )
+
+        # ?? conditional nightmare ?? refactor ??
```
Comment (FFroehlich) [inline python/sdist/amici/jax/petab.py]: 👍

File: python/sdist/amici/jax/petab.py [hunk 4]
```diff
@@ -709,15 +780,70 @@ def _eval_nn(self, output_par: str):
             .to_dict()
         )
 
+        condition_parameter_map = (
```
Comment (FFroehlich) [inline python/sdist/amici/jax/petab.py]: `condition_input_map`?

File: python/sdist/amici/jax/petab.py [hunk 5]
```diff
@@ -709,15 +780,70 @@ def _eval_nn(self, output_par: str):
...
+                    (
+                        petab_id,
+                        self._petab_problem.parameter_df.loc[
+                            self._petab_problem.condition_df.loc[
+                                condition_id, petab_id
+                            ],
+                            petab.NOMINAL_VALUE,
+                        ],
+                    )
+                    if "input"
```
Comment (FFroehlich) [inline python/sdist/amici/jax/petab.py]: this check seems a bit too unspecific. I think you want to construct a sequence of petab id's that are mapped to `$nnId.inputs{[$inputArgumentIndex]{[$inputIndex]}}`?

File: python/sdist/amici/jax/petab.py [hunk 6]
```diff
@@ -994,6 +1134,8 @@ def run_simulation(
         nps: jt.Float[jt.Array, "nt *nnp"],  # noqa: F821, F722
         mask_reinit: jt.Bool[jt.Array, "nx"],  # noqa: F821, F722
         x_reinit: jt.Float[jt.Array, "nx"],  # noqa: F821, F722
+        init_override: jt.Float[jt.Array, "np"], # ?? what do these annotations mean ??
```
Comment (FFroehlich) [inline python/sdist/amici/jax/petab.py]: type hints via https://docs.kidger.site/jaxtyping/, supposed to be checked via https://github.com/beartype/beartype but currently not functioning, see https://github.com/AMICI-dev/AMICI/issues/2849

I think the dimension should be `nx` not `np` here.

File: tests/sciml/test_sciml.py [hunk 7]
```diff
@@ -202,8 +194,16 @@ def test_ude(test):
 
     # llh
     if test in (
+        # ?? cases where nn part of observable formula ??
```
Comment (FFroehlich) [inline tests/sciml/test_sciml.py]: I think this might have been cases with neural network architecture we do not support

File: python/sdist/amici/jax/petab.py [hunk 8]
```diff
@@ -1590,3 +1615,52 @@ def petab_simulate(
             )
         dfs.append(df_sc)
     return pd.concat(dfs).sort_index()
+
+def apply_grad_filter(problem: JAXProblem,):
```
Comment (FFroehlich) [inline python/sdist/amici/jax/petab.py]: This is a great solution! Only thing that I am a bit worried about is that in the current implementation `stop_gradient` is only applied when calling `problem` methods in the context of `run_simulations`, which may lead to confusion when trying to compute gradients outside of that context. My interpretation of the petab problem definition is that setting `estimate=0` means that gradient computation is permanently disabled and we should apply `apply_grad_filter` during `JaxProblem` instantiation.

**Comment by FFroehlich** [comment]
just checking test failures:

- Notebook tests, also fails on the base branch albeit for a different reasons, but this is not related to changes here
- mac os, this looks like an issue with PRs from a fork
- doc tests: also problem in base branch, h5py is missing from doc requirements
- sbml jax: unrelated, also failing in base branch

File: python/sdist/amici/jax/petab.py (hunk 2 above)
Comment (BSnelling) [reply python/sdist/amici/jax/petab.py]: I tried casting to a numpy array but the array persisted as float32s. I've defined the `dtype` based on the current jax config settings which I think is better than hard coding.

File: python/sdist/amici/jax/petab.py (hunk 5 above)
Comment (BSnelling) [reply python/sdist/amici/jax/petab.py]: I've updated this.  The complication is that the values here could be pulled from a nominal value in the parameter table or from a value in the conditions table, depending on whether the id appears in the parameters table. That's my understanding anyway.

File: tests/sciml/test_sciml.py (hunk 7 above)
Comment (BSnelling) [reply tests/sciml/test_sciml.py]: Fair enough, they've all been implemented now in any case!

File: python/sdist/amici/jax/petab.py [hunk 9]
```diff
@@ -588,6 +647,40 @@ def _get_nominal_parameter_values(
                         model,
                         model_pars[net_id][layer_id][attribute],
                     )
+
+        # set inputs in the model if provided
+        if len(nn_input_arrays) > 0:
+            for net_id in model_pars:
+                for input in model.nns[net_id].inputs:
+                    input_array = dict(
```
Comment (FFroehlich) [inline python/sdist/amici/jax/petab.py]: bit unwieldy could we simplify this?

File: python/sdist/amici/jax/petab.py [hunk 10]
```diff
@@ -1118,6 +1302,38 @@ def run_simulations(
...
+        )
+        init_override = jnp.stack(
+            [
+                jnp.array(
+                    [
+                        self._eval_nn(
+                            self._parameter_mappings[sc].map_sim_var[p], sc
+                        )
+                        if p
+                        in set(self._parameter_mappings[sc].map_sim_var.keys())
```
Comment (FFroehlich) [inline python/sdist/amici/jax/petab.py:1314]: This isn't quite the check I would expect here (rather something with `condition_df`/`mapping_df`), but fine to keep as is for now as we will anyways have to change the whole parameter mapping setup for v2.

**Review by FFroehlich — APPROVED** [review]
